improve documentation (#58)
CodeIntelligenceTesting/cifuzz: README.md

# cifuzz

> > **_IMPORTANT:_** This project is under active development. Be aware that the behavior of the commands 

Reviewer: For better readability in terminals, I would prefer to wrap all READMEs on 72 characters.

Author: done

**cifuzz** is a CLI tool that helps you to integrate and run fuzzing 
based tests into your project.

## Getting started
If you are new to the world of fuzzing, we recommend you to take a 
look at our [Glossary](docs/Glossary.md).

### Installation

#### Building from Source 

**Prerequisites**
* [git](https://git-scm.com/)
* [make](https://www.gnu.org/software/make/)
* [cmake](https://cmake.org/)
* [clang / llvm](https://clang.llvm.org/get_started.html)
* [go >= 1.18](https://go.dev/doc/install)
* [libcap](https://man7.org/linux/man-pages/man3/libcap.3.html)

**Ubuntu / Debian**
```bash
sudo apt install git make cmake clang llvm golang-go libcap-dev 
```

**Arch**
```bash
sudo pacman -S git make cmake clang llvm go libcap
```

To build **cifuzz** from source you have to execute the following steps:
```bash
git clone https://github.com/CodeIntelligenceTesting/cifuzz.git
cd cifuzz
make test
make install
```
If everything went fine, you will find the newly created directory 
`~/cifuzz`. Do not forget to add `~/cifuzz/bin` to your `$PATH`.

### Setup / Create your first fuzz target
**cifuzz** commands will interactively guide you through the needed 

> options and shows additional next steps. You will also find a complete 
> list of the available commands with all supported options and 
> parameters by calling `cifuzz command --help` or 

Reviewer: ```suggestion
options and show next steps. You can find a complete 
list of the available commands with all supported options and 
parameters by calling `cifuzz command --help` or 
```

or 

```suggestion
options and show next steps. A complete 
list of the available commands with all supported options and 
parameters can be found by calling `cifuzz command --help` or 
```

Author: done

[here](https://github.com/CodeIntelligenceTesting/cifuzz/wiki/cifuzz).

1. To initialize your project with cifuzz just execute `cifuzz init` 
in the root directory of your project. This will create a file named 
`cifuzz.yaml` containing the needed configuration.

2. The next step is to create a fuzz target: `cifuzz create`. This 
will create a stub for your fuzz test (aka fuzz target), lets say it 
is called `my_fuzz_test.cpp`.

3. Edit `my_fuzz_test.cpp` so it actually calls the function you want 
to test with the input generated by the fuzzer. To learn more about 
writing fuzz tests you can take a look at our 
[Tutorial](docs/How-To-Write-A-Fuzz-Test.md) or one of the 
[example projects](examples).

> 4. Start the fuzzing by executing `cifuzz run my_fuzz_test`. 

Reviewer: I think there's a step missing here: The user needs to make sure that the fuzz target can be built, either via `cmake` or any other build system if they provide the `--build-command` flag to `cifuzz run`.

Reviewer: What do you mean with "make it more explicit on the 4."? I think the user should not run into failures when they follow these steps.

Author: > Start the fuzzing by executing cifuzz run my_fuzz_test. cifuzz now tries to build the fuzz test and starts a fuzzing run.

This is currently 4. Maybe changing it to something like this:

> Start the fuzzing by executing cifuzz run my_fuzz_test. cifuzz will build the fuzz test before executing it, so make sure it is compiling without errors (cifuzz will tell you there is a problem).

I would like to avoid adding another point because of the following reasons:
1. Without changing the fuzz target `cifuzz run` should just run fine
2. When changing code everyone knows that it has to be compilable and usually knows how to do it. Especially when we tell him to add the CMake commands when creating the target

Reviewer: > Without changing the fuzz target cifuzz run should just run fine

It doesn't though if all you do is follow the steps listed here, because cifuzz can't build the fuzz target. If the build system is cmake, you have to add the "add_fuzz_test" line printed by `cifuzz create` to the `CMakeLists.txt` in order for `cifuzz run` to work, and if the build system is unknown the user has to figure out on their own how to build the fuzz target. So IMO there is clearly a "configure your build system to build the fuzz target" step missing here (with a pointer to the output of `cifuzz create`).

Author: I expect the user to follow the instructions given in the command... if not we also have to point it out in step 1 for 
```
find_package(cifuzz)
enable_fuzz_testing()
```
as well and basically for all build systems / cases later on.
And I think we should not cover the "unknown/other" build system part for now, as we are currently focusing on the CMake integration. If we supporting other build system I would prefer linking to an additional page covering all the build system specific stuff.

I would be ok with writing something like "please follow the instructions printed out by the command" on Step 3.

Author: And we can also rework the steps after having actual user feedback.

Reviewer: > I would be ok with writing something like "please follow the instructions printed out by the command" on Step 3.

I would be fine with that
